feat(card): Card component

maulik-thaker1 opened the pull request:
## Summary

- Card components with Single slot
- Clickable and simple variant

## ADO Story 

https://dev.azure.com/Kyndryl/Shidoka%20-%20Bridge%20Design%20System/_workitems/edit/1589763

## Notes

- As per UX & UI discussion we (dev) need to provide card wrapper with single slot inside with Clickable variant.

## Checklist

- [x] Used Conventional Commit messages as outlined in the contributing guide.
  - [x] Noted breaking changes (if any).
- [x] Documented/updated all props, events, slots, parts, etc with JSDoc.
  - [x] Ran the `analyze` command to update Storybook docs.
- [x] Added/updated Stories with controls to cover all variants.
- [x] Added/updated a basic unit test, at least, for each component.
- [x] Added/updated the Figma link for the Story's Design tab.
- [x] Added any new component exports to the src/index.ts file

brian-patrick-3 commented on src/components/reusable/card/card.ts:
```diff
@@ -0,0 +1,145 @@
+  static override styles = CardScss;
+  /** Card width. Default 264px. */
+  @property({ type: Number })
+  cardWidth = 264;
```
Width should be dictated by the Grid, so the card component itself should always be 100% width.

brian-patrick-3 replied:
I think maybe instead of a width prop, set a CSS width on the `:host` selector. Then it can be overridden this way:
```css
kyn-card { width: 100%; }
```
It needs to be able to use percentage width, especially within a grid.

brian-patrick-3 commented on src/components/reusable/card/card.ts:
```diff
@@ -0,0 +1,145 @@
+    return html`${this.type === 'clickable'
+      ? html`<a
+          href=${this.cardLink}
+          target="_blank"
```
We shouldn't force new tab here. We also will need to add an event to capture link clicks.

brian-patrick-3 commented on src/components/reusable/card/card.scss:
```diff
@@ -0,0 +1,105 @@
+  //   box-shadow: 0px 1px 2px 0px rgba(61, 60, 60, 0.25);
+  outline: 1px solid transparent;
+  transition: outline 200ms ease-out;
+  &:hover {
+    outline: 2px solid var(--kd-color-border-ui-hover);
+  }
```
Non-clickable card should not have a hover effect.

brian-patrick-3 reviewed (commented):
Added a couple initial comments

maulik-thaker1 commented on src/components/reusable/card/card.ts:
```diff
@@ -0,0 +1,145 @@
+ * @slot card-thumbnail - Slot for card thumbnail image.
+ * @slot tags - Slot for the tag component.
+ * @slot card-links - Slot for card links.
+ * @slot unnamed - Slot for card body content.
```
Done

brian-patrick-3 reviewed (changes requested):
Here's a few initial observations:

1. Inside Grid story:
   1. `md` grid col size seems incorrect, doesn't fill the screen.
   2. Show cards with varying content, so we can see that the cards have matching heights.
   3. Set `height: 100%;` on `kyn-card` so they match in height.
   4. Set `height: 100%;` on `.card-wrapper` so it grows with the :host container.
2. The 1px `border` doesn't persist between "clickable" and "normal" types, so the card size changes and content shifts slightly when changing type.
3. Based on what we just discussed about notifications, we might want to make the `.card-wrapper` div customizable, so we can alter the shadow and background color. The easiest way would be to add a CSS `part` on this tag. The downside of `part` is it opens it up to any CSS customization.
4. I wonder if we could do anything to internalize some of the pattern CSS so it's easier for consuming applications to re-use. That might require sub-components though hmm. Just some food for thought.

maulik-thaker1 commented:
> Here's a few initial observations:
> 
> 1. Inside Grid story:
>    
>    1. `md` grid col size seems incorrect, doesn't fill the screen.
>    2. Show cards with varying content, so we can see that the cards have matching heights.
>    3. Set `height: 100%;` on `kyn-card` so they match in height.
>    4. Set `height: 100%;` on `.card-wrapper` so it grows with the :host container.
> 2. The 1px `border` doesn't persist between "clickable" and "normal" types, so the card size changes and content shifts slightly when changing type.
> 3. Based on what we just discussed about notifications, we might want to make the `.card-wrapper` div customizable, so we can alter the shadow and background color. The easiest way would be to add a CSS `part` on this tag. The downside of `part` is it opens it up to any CSS customization.
> 4. I wonder if we could do anything to internalize some of the pattern CSS so it's easier for consuming applications to re-use. That might require sub-components though hmm. Just some food for thought.

1. Inside Grid : 
    i. I changed and did this : class="kd-grid__col--sm-2 kd-grid__col--md-4 kd-grid__col--lg-3"
    ii. ok for this I need to set props inside `<sample-card-component>` right now I hard-coded Title, Subtitle and Description  inside `<sample-card-component>`
    iii. Set. it's already matching actual height of card
    iv. ok
2. yeah its `border` for `normal` and `outline` for `clickable`. I changed that and made `outline` for both
3. Need to think about that as right now as per UX there is white background for card and box-shadow for clickable card

brian-patrick-3 commented:
In the grid example, `kyn-card` still needs `height: 100%;` so they are all the same height.

brian-patrick-3 commented on src/components/reusable/card/card.scss:
```diff
@@ -0,0 +1,34 @@
+    background-color: var(--kd-color-background-ui-default);
+    outline: 1px solid var(--kd-color-border-ui);
+    border-radius: 8px;
+    box-shadow: 0px 1px 2px 0px rgba(61, 60, 60, 0.25);
```
Pull in [Elevation](https://kyndryl-design-system.github.io/shidoka-foundation/?path=/docs/foundation-elevation--docs) here, now that it's available.

maulik-thaker1 replied:
Can you give me an example how to use that ?

brian-patrick-3 commented on src/components/reusable/card/card.scss:
```diff
@@ -0,0 +1,34 @@
+    box-shadow: 0px 1px 2px 0px rgba(61, 60, 60, 0.25);
+    padding: 24px;
+    transition: outline 200ms ease-out;
+    &:hover {
+      outline: 2px solid var(--kd-color-border-ui-hover);
+    }
```
Hover needs Elevation level 2

maulik-thaker1 replied:
ok will use elevation here. Transition is there : ` transition: outline 200ms ease-out;`

maulik-thaker1 replied:
Hover need elevation level 1 only as per figma for clickable version.

brian-patrick-3 commented:
:tada: This PR is included in version 1.9.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@kyndryl-design-system/shidoka-applications/v/1.9.0)
- [GitHub release](https://github.com/kyndryl-design-system/shidoka-applications/releases/tag/v1.9.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
